add polling support

jluebbe opened the pull request:
This is based on the concept described in https://github.com/rauc/rauc/issues/1114.

Documentation:
- https://rauc--1672.org.readthedocs.build/en/1672/advanced.html#update-polling
- https://rauc--1672.org.readthedocs.build/en/1672/reference.html#polling-section

TODO:
- [x] https://github.com/rauc/rauc/pull/1654 needs to be merged first.
- [x] add test case for inhibit files
- [x] add test case for HTTP 204 status code

jluebbe committed:
utils: add helper to get a string list from a GKeyFile

The helper also optionally checks the strings against an allow-list.

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

jluebbe committed:
add a Poller D-Bus interface

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

jluebbe committed:
test/conftest: allow running the service with a polling speedup factor

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

jluebbe committed:
context: implement a polling speedup factor for testing

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

jluebbe committed:
install: remember if a slot or artifact was updated

This will allow us to determine if a reboot should be performed after an
automatic installation.

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

jluebbe committed:
implement polling in the service

This implements large parts of #1114 (Simple Update Server Support).

The polling functionality in RAUC allows a device to periodically check
for updates from a specified source. It fetches bundle manifests, checks
for available updates, and initiates installations if necessary. This
functionality is intended to automate the update process and ensure that
the system remains up-to-date with minimal manual intervention.

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

jluebbe committed:
docs: document polling functionality and the new D-Bus API

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

jluebbe committed:
add tests for polling

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

hnez commented:
From using the feature it does not look like there is a lot of debug output (by default default) about which candidate-/install-/reboot-criteria are met by a polled bundle and which are not.
I think that would be useful.

Something along the lines of:

> Not considered because of unmet criterium: …

jluebbe commented on src/poll.c:
```diff
@@ -0,0 +1,575 @@
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	const gchar *cmd = r_context()->config->poll_reboot_cmd;
+	g_assert(cmd);
```
Correct, the default is missing.

jluebbe replied:
This should be fixed.

jluebbe committed:
test/nginx_backend: add support for explicit HTTP response codes

This is useful for testing handling of aspects like 204 no content.

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

hnez commented on test/conftest.py:
```diff
@@ -147,6 +147,7 @@ def _have_faketime():
 
 
 no_service = pytest.mark.skipif(string_in_config_h("ENABLE_SERVICE 1"), reason="Have service")
+needs_service = pytest.mark.skipif(not string_in_config_h("ENABLE_SERVICE 1"), reason="Have service")
```
```suggestion
needs_service = pytest.mark.skipif(not string_in_config_h("ENABLE_SERVICE 1"), reason="Need service")
```

hnez commented on test/test_polling.py:
```diff
@@ -0,0 +1,352 @@
+
+
+def test_poll_only(create_system_files, system, http_server):
+    """Test if the info command sends custom headers correctly."""
```
I think the docstrings in this file are not up to date.

hnez commented on docs/advanced.rst:
```diff
@@ -508,6 +508,143 @@ Beside some standard information, like the *boot ID*, the system's *uptime* or
+.. code-block:: cfg
+   :emphasize-lines: 3-4
+
+   [poll]
+   source=https://example.com/stable/my_product.raucb
+   interval-sec=300
+   max-interval-sec=3600
+
+In this example:
+
+* RAUC polls the source every 5 minutes as long as no error occurs.
+* Every time a poll attempt fails, the interval is increased by 5 minutes, until
+  it reaches 1 hour.
+* If an attempt is successful again, the interval is reduced to 5 minutes.
```
The documentation should contain best practices. Polling every five minutes sounds excessive to me.

How about e.g.

```
   interval-sec=14400
   max-interval-sec=86400
```

e.g. four hours and one day respectively. Four hours still sounds a bit too frequent for my taste, but at least [hacker news will not make fun of it](https://news.ycombinator.com/item?id=43829006).

hnez commented on docs/reference.rst:
```diff
@@ -381,6 +381,95 @@ For more information about using the streaming support of RAUC, refer to
+
+.. _poll-candidate-criteria:
+
+``candidate-criteria`` (optional, default is ``higher-semver``)
+  Specifies the conditions under which a new bundle is considered a **candidate**
+  for updating.
+  These conditions do **not** automatically trigger the installation of the
+  update.
+
+  Supported values are:
+
+  * ``higher-semver``: The new bundle's version (interpreted as a Semantic
+    Version) is higher than the current system version.
+  * ``different-version``: The new bundle's version string differs from the
+    current system version (regardless of ordering).
```
I use the DBus poller `status` property to get the `compatible`, `version`, `manifest_hash` and `effective_url` of the remote bundle. I want these (and the other fields I do not use) to always be up to date with the results of the previous poll.

I did not yet fully grasp if that is the case, or if this information is only available if `candidate-criteria` matches.
Should I set `candidate-criteria=` (empty) in order to also see the remote bundle information via the `status` property if the semver of the current image and the remote bundle happen to match?

(The code in question, if you are interested https://github.com/linux-automation/tacd/pull/90/commits/0ab21a3c075fb0f1a9c2e97ebf78ea3c281d2866 https://github.com/linux-automation/tacd/pull/90/commits/911e4895dff86b2889dded82f039aaca55eb038c)

jluebbe commented:
@ejoerns @hnez Which clock base should we use for the `last-attempt-time` and `last-sucess-time` propertie (`CLOCK_MONOTONIC`, `CLOCK_BOOTTIME` or `CLOCK_REALTIME`)? As polling may happen before the clock is synchronized, `CLOCK_REALTIME` may be ambiguous. With `CLOCK_BOOTTIME`, we could calculate how much time passed since the last attempt.

jluebbe commented:
I've renamed the `[poll]` section to `[polling]` as discussed internally. The related function and variable names have been changed as well.

jluebbe commented:
The last force-push now also renames the new source files.

ejoerns commented on qemu-test-rauc-config:
```diff
@@ -13,6 +13,9 @@ events=all
 [keyring]
 path=ca.cert.pem
 
+[streaming]
```
A note in the commit message why this is required as preparation would be helpful.

ejoerns commented on src/polling.c:
```diff
@@ -0,0 +1,598 @@
+	}
+
+	if (!r_context()->system_version) {
+		g_error("failed to set up polling: system version not provided via system-info handler");
```
Since this is runtime input, using `g_error()` does not look appropriate here.

ejoerns commented on src/polling.c:
```diff
@@ -0,0 +1,598 @@
+{
+	GError *ierror = NULL;
+
+	g_return_val_if_fail(polling_source, FALSE);
```
Add a check for the GError, too

ejoerns commented on src/config_file.c:
```diff
@@ -668,6 +668,151 @@ static GHashTable *parse_artifact_repos(const char *filename, const char *data_d
+		return FALSE;
+	}
+
+	config->polling_source = key_file_consume_string(key_file, "polling", "source", NULL);
```
If 'polling_source' needs to be a URL anyway, maybe it might make sense to name it 'polling_url'?  This also reduces the potential confusion with the `GSource` `polling_source`.

jluebbe commented on src/service.c:
```diff
@@ -554,8 +560,6 @@ static void r_on_bus_acquired(GDBusConnection *connection,
 {
 	GError *ierror = NULL;
 
-	r_installer = r_installer_skeleton_new();
```
It's required. I've split it into a separate commit.

jluebbe committed:
src/service: remember whether the booted slot is marked good

This will be used to inhibit polling until we know that we have a
good slot to fall back to.

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>

jluebbe commented on src/polling.c:
```diff
@@ -0,0 +1,598 @@
+	return G_SOURCE_REMOVE;
+}
+
+static gboolean polling_install(RPollingSource *polling_source, GError **error)
```
I've removed the `error` parameter.

jluebbe commented on src/install.c:
```diff
@@ -48,6 +48,10 @@ static void install_args_update(RaucInstallArgs *args, const gchar *msg, ...)
 	g_return_if_fail(args);
 	g_return_if_fail(msg);
 
+	/* without a notify function, there is nothing to do */
```
I've extended the comment.

ejoerns commented on docs/advanced.rst:
```diff
@@ -547,6 +508,183 @@ Beside some standard information, like the *boot ID*, the system's *uptime* or
+
+To ensure that the correct bundle is installed after evaluating the manifest
+information in your application, pass the *manifest hash* back to RAUC by
+calling the InstallBundle method with the ``require-manifest-hash`` argument.
```
This could link to `_gdbus-method-de-pengutronix-rauc-Installer.InstallBundle` (and maybe note what happens if the provided hash does not match the expected one.
